api/osb/osb_controller.go

```diff
@@ -35,42 +37,97 @@ import (
 
 var osbPathPattern = regexp.MustCompile("^" + web.OSBURL + "/[^/]+(/.*)$")
 
-// BrokerFetcher is implemented by OSB handler providers
+// BrokerFetcher is implemented by OSB proxy providers
 type BrokerFetcher interface {
 	FetchBroker(ctx context.Context, brokerID string) (*types.Broker, error)
 }
 
+// CatalogFetcher is implemented by OSB catalog providers
+type CatalogFetcher interface {
+	FetchCatalog(ctx context.Context, brokerID string) (*types.ServiceOfferings, error)
+}
+
+// CatalogFetcherFunc allows functions to implement CatalogFetcher
+type CatalogFetcherFunc func(ctx context.Context, brokerID string) (*types.ServiceOfferings, error)
+
+// FetchCatalog implements osb.CatalogFetcher and uses the provided function to fetch the catalog for the specified broker
+func (rhf CatalogFetcherFunc) FetchCatalog(ctx context.Context, brokerID string) (*types.ServiceOfferings, error) {
+	return rhf(ctx, brokerID)
+}
+
 // controller implements api.Controller by providing OSB API logic
 type controller struct {
-	fetcher BrokerFetcher
+	brokerFetcher  BrokerFetcher
+	catalogFetcher CatalogFetcher
 }
 
 var _ web.Controller = &controller{}
 
 // NewController returns new OSB controller
-func NewController(fetcher BrokerFetcher, _ http.RoundTripper) web.Controller {
-	return &controller{
-		fetcher: fetcher,
+func NewController(brokerFetcher BrokerFetcher, catalogFetcher CatalogFetcher, _ http.RoundTripper) web.Controller {
+	controller := &controller{
+		brokerFetcher:  brokerFetcher,
+		catalogFetcher: catalogFetcher,
 	}
+	return controller
+}
+
+func (c *controller) proxyHandler(r *web.Request) (*web.Response, error) {
+	return c.handler(r, c.proxy)(r)
+}
+
+func (c *controller) catalogHandler(r *web.Request) (*web.Response, error) {
+	return c.handler(r, c.catalog)(r)
 }
 
-func (c *controller) handler(r *web.Request) (*web.Response, error) {
+func (c *controller) handler(r *web.Request, f func(r *web.Request, logger *logrus.Entry, id string) (*web.Response, error)) func(request *web.Request) (*web.Response, error) {
```

Comment:
Why does this **handler(...)** function receive a *web.Request* argument?
Shouldn't the below returned function make use of the **request** argument rather than this **r** variable?